scrape/scrape.go

```diff
@@ -932,77 +932,99 @@ mainLoop:
 		default:
 		}
 
-		var (
-			start             = time.Now()
-			scrapeCtx, cancel = context.WithTimeout(sl.ctx, timeout)
-		)
+		last = sl.scrapeAndReport(interval, timeout, last, errc)
 
-		// Only record after the first scrape.
-		if !last.IsZero() {
-			targetIntervalLength.WithLabelValues(interval.String()).Observe(
-				time.Since(last).Seconds(),
-			)
+		select {
+		case <-sl.parentCtx.Done():
+			close(sl.stopped)
+			return
+		case <-sl.ctx.Done():
+			break mainLoop
+		case <-ticker.C:
 		}
+	}
 
-		b := sl.buffers.Get(sl.lastScrapeSize).([]byte)
-		buf := bytes.NewBuffer(b)
+	close(sl.stopped)
 
-		contentType, scrapeErr := sl.scraper.scrape(scrapeCtx, buf)
-		cancel()
+	if !sl.disabledEndOfRunStalenessMarkers {
+		sl.endOfRunStaleness(last, ticker, interval)
+	}
+}
 
-		if scrapeErr == nil {
-			b = buf.Bytes()
-			// NOTE: There were issues with misbehaving clients in the past
-			// that occasionally returned empty results. We don't want those
-			// to falsely reset our buffer size.
-			if len(b) > 0 {
-				sl.lastScrapeSize = len(b)
-			}
-		} else {
-			level.Debug(sl.l).Log("msg", "Scrape failed", "err", scrapeErr.Error())
-			if errc != nil {
-				errc <- scrapeErr
-			}
-		}
+// scrapeAndReport performs a scrape and then appends the result to the storage
+// together with reporting metrics, all using the same appender.
```

Comment:
"the same" is a little misleading as a 2nd appender will be created if there's a scrape error.

Reply:
But in case of a scrape error, the 1st appender is rolled back, so it will not append anything to the storage, so only one and the same appender will ever append any samples to the storage.

But if you have a better suggestion for the wording, I'm sure @roidelapluie will appreciate it.

Reply:
// scrapeAndReport performs a scrape and then appends the result to the storage
// together with reporting metrics, by using as few appenders as possible.
// In the happy scenario, a single appender is used.